internal/directives/kustomize_set_image_directive.go

```diff
@@ -0,0 +1,379 @@
+package directives
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/xeipuuv/gojsonschema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kustomize/api/konfig"
+	kustypes "sigs.k8s.io/kustomize/api/types"
+	kyaml "sigs.k8s.io/kustomize/kyaml/yaml"
+	yaml "sigs.k8s.io/yaml/goyaml.v3"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	intyaml "github.com/akuity/kargo/internal/yaml"
+)
+
+// preserveSeparator is the separator used to preserve values in the
+// Kustomization image field.
+const preserveSeparator = "*"
+
+func init() {
+	// Register the kustomize-set-image directive with the builtins registry.
+	builtins.RegisterDirective(
+		newKustomizeSetImageDirective(),
+		&DirectivePermissions{
+			AllowKargoClient: true,
+		},
+	)
+}
+
+// kustomizeSetImageDirective is a directive that sets images in a Kustomization
+// file.
+type kustomizeSetImageDirective struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newKustomizeSetImageDirective creates a new kustomize-set-image directive.
+func newKustomizeSetImageDirective() Directive {
+	return &kustomizeSetImageDirective{
+		schemaLoader: getConfigSchemaLoader("kustomize-set-image"),
+	}
+}
+
+func (d *kustomizeSetImageDirective) Name() string {
+	return "kustomize-set-image"
+}
+
+func (d *kustomizeSetImageDirective) Run(ctx context.Context, stepCtx *StepContext) (Result, error) {
+	// Validate the configuration against the JSON Schema.
+	if err := validate(d.schemaLoader, gojsonschema.NewGoLoader(stepCtx.Config), d.Name()); err != nil {
+		return ResultFailure, err
+	}
+
+	// Convert the configuration into a typed object.
+	cfg, err := configToStruct[KustomizeSetImageConfig](stepCtx.Config)
+	if err != nil {
+		return ResultFailure,
+			fmt.Errorf("could not convert config into kustomize-set-image config: %w", err)
+	}
+
+	return d.run(ctx, stepCtx, cfg)
+}
+
+func (d *kustomizeSetImageDirective) run(
+	ctx context.Context,
+	stepCtx *StepContext,
+	cfg KustomizeSetImageConfig,
+) (Result, error) {
+	// Find the Kustomization file.
+	kusPath, err := findKustomization(stepCtx.WorkDir, cfg.Path)
+	if err != nil {
+		return ResultFailure, fmt.Errorf("could not discover kustomization file: %w", err)
+	}
+
+	// Discover image origins and collect target images.
+	images, err := discoverImages(ctx, stepCtx.KargoClient, stepCtx.Project, cfg.Images, stepCtx.FreightRequests)
```

Comment:
🤔  The existing `freight.FindImage(...)` func should be reusable as is.

I'm currently using its buddy `freight.FindCommit(...)` in #5.

Reply:
The `Image` in this scenario is different from the `Image` in the Kargo API. They look alike and are partially, but in the broader sense, they are quite different.

Reply:
Gotcha. I'll look closer. But is this bit of discovery not digging through a Freight collection and, if necessary, attempting to discern the desired origin just-in-time?

Regardless... I'll take that closer look.

Reply:
Adding that I didn't have any significant feedback otherwise. I admittedly glossed over some of the low-level YAML stuff, but I trust that your tests cover it well.

Reply:
Made the desired changes in ca7cffda429c39f94fef0a5182e4f35c077dfb79, I think.